### PR TITLE
Introduce boolean expressions

### PR DESCRIPTION
**Pull request by felix-oq**
Preparation for #213

Allows users to use boolean expressions wherever a boolean value is expected. Introduces the following operators (ordered descending by [precedence](https://en.wikipedia.org/wiki/Order_of_operations)):

- `not`: unary operator to negate a boolean value
- `==` / `!=`: binary operator to check (in)equality
- `xor`: binary operator to compute exclusive OR
- `and`: binary operator to compute AND
- `or`: binary operator to compute OR

Example for a boolean expression: `(not true or false) xor true and true` which evaluates to `true`.

Also adds a diagnostic to propose a simplification for expressions that can be evaluated at compile-time:
image

### CODE REVIEW

**Review by georg-schwarz — APPROVED** [review]
Awesome!

File: libs/language-server/src/lib/ast/model-util.ts [hunk 1]
```diff
@@ -209,15 +216,68 @@ export function inferTypesFromValue(
...
+        return !evaluateExpression(expression.expression);
+      }
+      default:
+        assertUnreachable(unaryOperator);
+    }
+  }
+  if (isBinaryExpression(expression)) {
+    const binaryOperator = expression.operator;
+    switch (binaryOperator) {
+      case '==': {
```
Comment (rhazn) [inline libs/language-server/src/lib/ast/model-util.ts:243]: Imho these should be constants, not strings.

**Review by rhazn — APPROVED** [review]

File: libs/language-server/src/lib/ast/model-util.ts (hunk 1 above)
Comment (felix-oq) [reply libs/language-server/src/lib/ast/model-util.ts:243]: Note that they are not hardcoded. Instead, they are compliant with the literal type of the generated AST.

Excerpt from the generated `ast.ts`:
``` typescript
export interface BinaryExpression extends AstNode {
    readonly $container: BinaryExpression | CollectionLiteral | PropertyAssignment | UnaryExpression | ValuetypeDefinition;
    readonly $type: 'BinaryExpression';
    left: BooleanExpression
    operator: '!=' | '==' | 'and' | 'or' | 'xor'
    right: BooleanExpression
}
```

Comment (rhazn) [reply libs/language-server/src/lib/ast/model-util.ts:243]: Not really the only thing enums would be good for but in that case I don't feel bad about doing it like this, makes sense 👍

**Pull request merged**